src/protocol-muxer.js

```diff
@@ -1,21 +1,29 @@
 'use strict'
 
 const multistream = require('multistream-select')
+const observeConn = require('./observe-connection')
 
-module.exports = function protocolMuxer (protocols, conn) {
-  const ms = new multistream.Listener()
+module.exports = function protocolMuxer (protocols, observer) {
+  return (transport) => (parentConn) => {
+    const ms = new multistream.Listener()
 
-  Object.keys(protocols).forEach((protocol) => {
-    if (!protocol) {
-      return
-    }
+    Object.keys(protocols).forEach((protocol) => {
+      if (!protocol) {
+        return
+      }
```

Comment:
Still digesting the code but it seems that there is an assumption that all the switch.handle happen in the beginning and so you are locking the list of protocols we support.

Reply:
The protocols is an object, which can get mutated later as the user adds more protocols.

Reply:
Let's not get https://github.com/libp2p/js-libp2p-switch/issues/239 even harder to fix :)

Reply:
@diasdavid again, this was the same behaviour as before: `switch.protocols` also contained a mutable object with the protocol handlers... I'm just storing the protocols on that closure, but it's still a JS object that will get mutated, same as before..

Reply:
Got it. Makes sense now thanks for explaining :)